FBRetainCycleDetector/Detector/FBRetainCycleDetector.mm

```diff
@@ -112,31 +112,33 @@ - (void)addCandidate:(id)candidate
     @autoreleasepool {
       // Take topmost node in stack and mark it as visited
       FBNodeEnumerator *top = [stack lastObject];
+      // Take object that topmost node monitor
+      __weak id object = top.object.object;
 
       // We don't want to retraverse the same subtree
-      if (![objectsOnPath containsObject:top]) {
-        if ([_objectSet containsObject:@([top.object objectAddress])]) {
+      if (![objectsOnPath containsObject:object]) {
+        if ([_objectSet containsObject:object]) {
           [stack removeLastObject];
           continue;
         }
-        // Add the object address to the set as an NSNumber to avoid
-        // unnecessarily retaining the object
-        [_objectSet addObject:@([top.object objectAddress])];
+
+        [_objectSet addObject:object];
       }
 
-      [objectsOnPath addObject:top];
+      [objectsOnPath addObject:object];
 
       // Take next adjecent node to that child. Wrapper object can
       // persist iteration state. If we see that node again, it will
       // give us new adjacent node unless it runs out of them
       FBNodeEnumerator *firstAdjacent = [top nextObject];
-      if (firstAdjacent) {
+      __weak id firstAdjacentObject = firstAdjacent.object.object;
```

Comment:
Maybe it needs to mimic [`storeWeak`](https://github.com/zhongwuzw/objc4-cn/blob/master/runtime/objc-weak.mm#L393-L418)?